crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,227 @@
+//! struct type names are expressed in camel case).  In other words, if a module defines a struct
+//! type whose name is the same as the module name, this type MUST possess these special properties,
+//! otherwise the module definition will be considered invalid and will be rejected by the
+//! validator:
+//!
+//! - it has only one ability: drop
+//! - it has only one arbitrarily named field of type boolean (since Move structs cannot be empty)
+//! - its definition does not involve type parameters
+//! - its only instance in existence is passed as an argument to the module initializer
+//! - it is never instantiated anywhere in its defining module
+
+use move_binary_format::{
+    access::ModuleAccess,
+    binary_views::BinaryIndexedView,
+    file_format::{
+        Ability, AbilitySet, Bytecode, CompiledModule, FunctionDefinition, FunctionHandle,
+        SignatureToken, StructDefinition, StructHandle, TypeSignature,
+    },
+};
+use move_core_types::{ident_str, language_storage::ModuleId};
+use sui_types::{
+    base_types::{TX_CONTEXT_MODULE_NAME, TX_CONTEXT_STRUCT_NAME},
+    error::ExecutionError,
+    SUI_FRAMEWORK_ADDRESS,
+};
+
+use crate::{verification_failure, INIT_FN_NAME};
+
+pub fn verify_module(module: &CompiledModule) -> Result<(), ExecutionError> {
+    // In Sui's framework code there is an exception to the characteristic type rule - we have a SUI
+    // type in the sui module but it is instantiated and the module has no initializer (the reason
+    // for it is that the SUI coin is only instantiated during genesis). It is easiest to simply
+    // special-case this module particularly that this is framework code and thus deemed correct.
+    if ModuleId::new(SUI_FRAMEWORK_ADDRESS, ident_str!("sui").to_owned()) == module.self_id() {
+        return Ok(());
+    }
+
+    let view = BinaryIndexedView::Module(module);
+    let mod_handle = view.module_handle_at(module.self_module_handle_idx);
+    let mod_name = view.identifier_at(mod_handle.name).as_str();
+    let struct_defs = &module.struct_defs;
+    let mut char_type_candidate = None;
+    for def in struct_defs {
+        let struct_handle = module.struct_handle_at(def.struct_handle);
+        let struct_name = view.identifier_at(struct_handle.name).as_str();
+        if struct_name.to_ascii_uppercase() == struct_name
+            && mod_name.to_ascii_uppercase() == struct_name
+        {
+            verify_char_type(module, struct_name, struct_handle, def)
+                .map_err(verification_failure)?;
+            // if we reached this point, it means we have a legitimate characteristic type candidate
+            // and we have to make sure that both the init function's signature reflects this and
+            // that this type is not instantiated in any function of the module
+            char_type_candidate = Some((struct_name, struct_handle, def));
+            break; // no reason to look any further
+        }
+    }
```

Comment:
All this would be a bit cleaner with `struct_defs.iter().find`

Reply:
Possibly, but then we'd have to deal with a return value being a vector, explain why it would only have one or zero elements, etc. Overall, I am not sure it's a huge win, but happy to do the rewrite if requested one more time :-)

Reply:
`find` gives the first element that satisfies the predicate, so what you have here is equivalent to `find`, no vectors or anything

Reply:
I didn't realize the `Result` in here. Really I think you need `try_find`, which is still in nightly